packages/polkadart_scale_codec/lib/src/core/codec.dart

```diff
@@ -453,18 +453,84 @@ class Codec {
 
   ///
   /// Decodes Bit Sequence
-  List<int> _decodeBitSequence(Source source) {
-    var length = (source.compactLength() / 8).ceil();
-    return source.bytes(length).toList();
+  List<int> _decodeBitSequence(BitSequenceType def, Source source) {
+    // get compacted length for filling result
+    final int length = source.compactLength();
+
+    // get the capacity of the bit sequence
+    final int capacity = max(length, _getBitSequenceCapacity(def.bitStoreType));
+
+    // get the encoded bytes from the source
+    final List<int> bytes = source.bytes((capacity / 8).ceil()).toList();
+
+    // get the extra bit length so to form a 4 * Uint8List
+    final int extraBitLength = capacity - bytes.length;
+
+    // add the 0 to the bits
+    for (var i = 0; i < extraBitLength; i++) {
+      bytes.add(0);
+    }
+
+    final BitArray bitArray = BitArray.fromUint8List(Uint8List.fromList(bytes));
+
+    final List<int> bits = List.filled(length, 0);
+
+    for (var i = 0; i < length; i++) {
+      bits[i] = bitArray[i] ? 1 : 0;
+    }
+
+    return bits;
   }
 
   ///
   /// Encodes Bit Sequence
-  void _encodeBitSequence(dynamic bits, ScaleCodecEncoder encoder) {
-    assertionCheck(
-        bits is List<int>, 'BitSequence can have bits of type List<int> only.');
-    encoder.compact(bits.length * 8);
-    encoder.bytes(bits);
+  void _encodeBitSequence(
+      BitSequenceType def, dynamic bits, ScaleCodecEncoder encoder) {
+    assertionCheck(bits is List<int>,
+        'BitSequence can have bits of type `List<int>` only.');
+
+    int capacity = max(bits.length, _getBitSequenceCapacity(def.bitStoreType));
+
+    capacity = (capacity / 8).ceil() * 8;
+
+    encoder.compact(bits.length);
+
+    final bitArray = BitArray(capacity);
+    for (var i = 0; i < bits.length; i++) {
+      switch (bits[i]) {
+        case 0:
+          bitArray[i] = false;
+          break;
+        case 1:
+          bitArray[i] = true;
+          break;
+        default:
+          throw UnexpectedCaseException(
+              'Invalid byte, unable to encode $bits.');
+      }
+    }
+
+    encoder.bytes(
+        bitArray.byteBuffer.asUint8List(0, (capacity / 8).ceil()).toList());
+  }
```

Comment:
@justkawal this is what I mean by create a new variable - also doing some clean up
```suggestion
    final int capacity = max(bits.length, _getBitSequenceCapacity(def.bitStoreType));
    final int byteCapacity =  (capacity / 8).ceil()
    final bitCapacity = byteCapacity * 8;
    final bitArray = BitArray(bitCapacity);

    encoder.compact(bits.length);

    for (var i = 0; i < bits.length; i++) {
      if(i > 1 || i < 0) { continue; } // Not sure if this is necessary
      bitArray[i] = i.isOdd;
    }
    encoder.bytes(bitArray.byteBuffer.asUint8List(0, byteCapacity).toList());
  }
```